### ISSUE TITLE
Sample window font in Firefox (macOS) is blurry

### ISSUE THREAD

**Issue by JCash** [Celtoys/Remotery#226]
<img width="177" alt="Screenshot 2022-07-20 at 09 46 07" src="https://user-images.githubusercontent.com/1349334/179935396-85d3883a-0c9e-495d-87ee-0648a4c1bb05.png">
<img width="699" alt="Screenshot 2022-07-20 at 10 29 56" src="https://user-images.githubusercontent.com/1349334/179935581-a3a48fe4-7174-49d2-ab1e-14e5fae83ebb.png">

The `W` and `S` are most visibly cut off.

I tried to verify the texture with the Spector.js extension, but it wouldn't run for me.
Although I expect it to be the shader code to be slightly off here.

**Comment by dwilliamson** [comment]
Yes, there's not much I can do about this at the moment beyond making the font and spacing bigger, consuming more space in the UI.

The font renderer needs to be a lot more sophisticated to match the browser.

**Comment by JCash** [comment]
What's confusing to me is that the header font looks a lot more crisp.
Is that the same font? The S's look diffferent to me, so I guess not?

**Comment by dwilliamson** [comment]
I think an immediate improvement could be had with a pre-generated font atlas.

Can you run this for me: https://lucide.github.io/Font-Atlas-Generator/

Use these properties:
* Font: Fira Code
* Font Size: 9

Do the letters in the atlas look better than they do in Remotery?

**Comment by dwilliamson** [comment]
In fact, it appears they have the same problem in that generator: you can't reliably generate letters as the amount of anti-aliasing varies depending on where the letter is rendered into the atlas. These are two of the sample letter positioned differently on the atlas:

![image](https://user-images.githubusercontent.com/1532903/180086017-c5e4dc9d-e1a7-456c-9aa8-a67624c1bc77.png)
![image](https://user-images.githubusercontent.com/1532903/180086052-096c4d1b-2b8d-4231-9c91-c34dd5a3bc04.png)

What also seems to vary it is the width of the output rectangle: odd generates different to even.

I think for reliability for now, I'll need a custom-built atlas. Then later (much later) quality font rendering can be addressed.

**Comment by dwilliamson** [comment]
I might try this if I get time https://github.com/memononen/fontstash

**Comment by JCash** [comment]
Sorry, I couldn't get the font atlas generator to produce a Fira Code atlas (or Fira Sans of any kind).

Yes, if the atlas is problematic to start with, it sounds good to look into alternatives.

**Comment by dwilliamson** [comment]
Linking for reference https://www.warp.dev/blog/adventures-text-rendering-kerning-glyph-atlases

**Comment by dwilliamson** [comment]
Rather than generating the atlas offline, we might find some clues on water-tight atlas generation from xterm.js: 

* https://github.com/xtermjs/xterm.js/tree/master/addons/xterm-addon-webgl/src
* https://github.com/xtermjs/xterm.js/pull/1790

Interestingly, it appears to render a quad per glyph, so is still quite a ways from being the most optimal. This may imply requirements of their atlas generation that we can't match.
